sycl/include/CL/sycl/ONEAPI/group_algorithm.hpp

```diff
@@ -190,6 +190,28 @@ using EnableIfIsNonNativeOp = cl::sycl::detail::enable_if_t<
         !cl::sycl::detail::is_native_op<T, BinaryOperation>::value,
     T>;
 
+template <typename T>
+struct is_native_function_object_impl : std::false_type {};
+
+template <template <typename> class BinaryOperation, typename T>
+struct is_native_function_object_impl<BinaryOperation<T>>
+    : std::integral_constant<
+          bool, (sycl::detail::is_scalar_arithmetic<T>::value ||
+                 sycl::detail::is_vector_arithmetic<T>::value ||
+                 std::is_same<T, void>::value) &&
```

Comment:
In case of T is void (when real type will be auto-deduced  on binary_op call) - how an identity will be auto-detected?
In general - it is not possible...

Reply:
I agree that's the interesting question

Reply:
An identity value isn't always required by the group algorithms.  For example, `reduce` and `exclusive_scan` all take an `init` parameter for the cases where the identity value is unknown.

We should introduce a separate type trait for whether a functor has a known identity, and it shouldn't be tied to whether the functor is declared in the standard.  @v-klochkov's already has something like this in the `reduction` implementation; we need to make it official, and provide a mechanism for users to declare the identity values for custom functors.

Reply:
> An identity value isn't always required by the group algorithms. For example, reduce and exclusive_scan all take an init parameter for the cases where the identity value is unknown.

John, it seems that we may incorrect result in case of considering an init parameter instead of identity... Or you want to
say  the group algorithms don't  combine the partial "sums"? ('m not properly aware about the group algorithms. Please, correct  me if I mistake.)

Reply:
> John, it seems that we may incorrect result in case of considering an init parameter instead of identity... Or you want to
say the group algorithms don't combine the partial "sums"? ('m not properly aware about the group algorithms. Please, correct me if I mistake.)

The kernel-wide `reduction` needs an identity value for optimization purposes, because it's possible for a work-item executing a kernel to contribute nothing to the reduction.  In such cases, the fact that the work-item didn't contribute anything needs to be tracked somehow (e.g. by storing a `std::optional` or an extra `bool`), or the work-item's contribution can be initialized to a value that doesn't impact the result (i.e. the identity value).

With the most common forms of `reduce` and `exclusive_scan`, all work-items contribute a value to the operation, and so that initialization issue goes away.  I think the only form that strictly requires a known identity is `sycl::exclusive_scan(Group, T, BinaryOperation)`, because it's not obvious what value to return on the first work-item otherwise.